governance/program/src/state/proposal.rs

```diff
@@ -102,11 +110,182 @@ impl Proposal {
+        if self
+            .voting_at
+            .unwrap()
+            .checked_add(config.max_voting_time)
+            .unwrap()
+            < current_slot
+        {
+            return Err(GovernanceError::ProposalVotingTimeExpired.into());
+        }
+
+        Ok(())
+    }
+
+    /// Checks if Proposal can be finalized
+    pub fn assert_can_finalize_vote(
+        &self,
+        config: &GovernanceConfig,
+        current_slot: Slot,
+    ) -> Result<(), ProgramError> {
+        if self.state != ProposalState::Voting {
+            return Err(GovernanceError::InvalidStateCannotFinalize.into());
+        }
+
+        // Check if we passed the configured max_voting_time period yet
+        if self
+            .voting_at
+            .unwrap()
+            .checked_add(config.max_voting_time)
+            .unwrap()
+            >= current_slot
+        {
+            return Err(GovernanceError::CannotFinalizeVotingInProgress.into());
+        }
+
+        Ok(())
+    }
+
+    /// Finalizes vote by moving it to final state Succeeded or Defeated if max_voting_time has passed
+    /// If Proposal is still within max_voting_time period then error is returned
+    pub fn finalize_vote(
+        &mut self,
+        governing_token_supply: u64,
+        config: &GovernanceConfig,
+        current_slot: Slot,
+    ) -> Result<(), ProgramError> {
+        self.assert_can_finalize_vote(config, current_slot)?;
+
+        let current_yes_vote_threshold =
+            self.get_current_vote_threshold(self.yes_votes_count, governing_token_supply);
+
+        let final_state = if current_yes_vote_threshold > config.vote_threshold_percentage as u64 {
+            ProposalState::Succeeded
+        } else {
+            ProposalState::Defeated
+        };
+
+        self.state = final_state;
+        self.voting_completed_at = Some(current_slot);
+
+        Ok(())
+    }
+
+    /// Returns current vote threshold in relation to the total governing token supply
+    fn get_current_vote_threshold(&self, vote_count: u64, governing_token_supply: u64) -> u64 {
+        vote_count
+            .checked_mul(100)
+            .unwrap()
+            .checked_div(governing_token_supply)
+            .unwrap()
+    }
+
+    /// Checks if vote can be tipped and automatically transitioned to Succeeded or Defeated state
+    /// If the conditions are met the state is updated accordingly
+    pub fn try_tip_vote(
+        &mut self,
+        governing_token_supply: u64,
+        config: &GovernanceConfig,
+        current_slot: Slot,
+    ) -> Result<(), ProgramError> {
+        if let Some(tipped_state) = self.try_get_tipped_vote_state(governing_token_supply, config) {
+            self.state = tipped_state;
+            self.voting_completed_at = Some(current_slot);
+        }
+
+        Ok(())
+    }
+
+    /// Checks if vote can be tipped and automatically transitioned to Succeeded or Defeated state
```

Comment:
Yeah, I've been contemplating this idea myself but eventually decided this should not be allowed for the following reasons

- As you I thought it could drive community engagement and result in quicker votes
- It could be confusing because the final vote tally could be different from the actual one. For example vote could  pass with  35% Yes, 20% No for 30% threshold, but then users could add more No votes. 
- With casting votes on already completed Vote there is an issue with your staked tokens. You can't withdraw tokens once vote is pending, but after that you can. If you casted your vote on completed proposals we would have to hold your tokens indefinitely.

I can think about the following solutions if we want all community opinion:
- Add an option to like/dislike completed votes 
- Add an option to disable auto tipping and always require manual `FinaliseVote`

Reply:
That all makes sense to me, glad to hear we're on the same page :smile:

Reply:
:)